Add symbols for this, super to operation constraint scope

fxjordan opened the pull request:
## What?
Add the following symbols to the `OCLOperationConstraint` scope
- a VariableSymbol `this` enabling access to the declaring type if the method
- if the declaring type has a super type, add `super` VariableSymbol (same as for `OCLInvariant` context definition)
- ~all `FunctionSymbol`s of the declaring type~
- only add VariableSymbols enclosed in the the method declaring type if they do not conflict with another VariableSymbol (usually a method parameter)

## Why?
- We already had access to fields of the declaring type but not to the methods. Using query methods (not modifying the system state) is allowed in OCL
- To avoid confusion, why not add `this` and `super` as in invariants?

## Tests
I would have added test cases covering this, but the test directory structure is super confusing without explanation. Where can I add a simple test case covering that the symbols can be used and symbol table creation is sucessful? @MaxStachon

fxjordan committed:
feat: add symbols for this,super available functions to operation constraint scope

fxjordan commented:
Any objections @MaxStachon ? ~Is there a reason these symbols have not been added yet?~ 

**Update**: I already realized there is a type check issue when a VariableSymbol and FunctionSymbol have the same name. So functions have to be accessed using `this.myFunction()`

fxjordan committed:
fix: add VariableSymbols only if there is not a local symbol yet(method parameter)
fix: revert adding FunctionSymbols

SE-FDr commented:
1. The TypeCheck should be able to handle variables and functions with the same name? What is the error?
2. Tests are missing!
3. Not every type with Supertypes has a SuperClass.

fxjordan commented:
Thanks your the fedback @SE-FDr 

> 1. The TypeCheck should be able to handle variables and functions with the same name? What is the error?

I have the following example
```java
ocl BuilderRef {

	context EmployeeBuilder EmployeeBuilder.firstName(String firstName)
	    post: build().firstName == firstName;
}

classdiagram BuilderRef {

  class Employee {
    String firstName;
  }

  class EmployeeBuilder {
    String firstName;
    EmployeeBuilder firstName(String firstName);
    Employee build();
  }
}
```
resulting in this error:
```
[ERROR]  MethodCallChangesBuildResultOut.ocl:<7,10> - MethodCallChangesBuildResultOut.ocl:<7,40>: 0xB0166 Operator '==' not applicable to 'String', '(String -> EmployeeBuilder) & String'
[ERROR]  0xOCL07 type of postcondition expression build().firstName==firstName could not be calculated.
```
> 2. Tests are missing!

Sure, I'll add some. I was just unsure about the directory. Is `testInput/parsable/symtab/coco/not_javagen` the correct one for this PR?

> 3. Not every type with Supertypes has a SuperClass.

True. I copied this from `visit(ASTOCLInvariant` which has the same issue. I guess checking `isPresentSuperClass()` is more precise.

SE-FDr commented:
1. This is the expected behavior.
You are comparing `firstName`, but you are not specifying _which_ `firstName`; do you want to compare the variable or the function? -> Thus, the expression has unknown semantics.
(yes, functions can be used without calling them, functions are first class citizens)
2. Yes, seems good
3. Yes, that would be great.

SE-FDr commented:
Attendum:
one could select the variable by casting, e.g., 
`post: ((String)build().firstName) == ((String)firstName);`
as this forces the selection of the variable
(I did not verify this, and yes, I am aware that this is ugly (but I am not aware of a better solution to make it semantically correct in all cases otherwise))

fxjordan commented:
> Attendum: one could select the variable by casting, e.g., `post: ((String)build().firstName) == ((String)firstName);` as this forces the selection of the variable (I did not verify this, and yes, I am aware that this is ugly (but I am not aware of a better solution to make it semantically correct in all cases otherwise))

I tried to test it, but the type cast expression is part of the UglyExprssions language which is not extended by OCL. So, maybe, not adding function to the scope is the best workaround we have at the moment.

fxjordan committed:
fix: check isPresentSuperClass instead of isEmptySuperTypes

fxjordan committed:
fix: deepClone VariableSymbol before adding it to scope of OperationConstraint

fxjordan committed:
test: add tests covering this/super usage in invariant/operation constraint

fxjordan commented:
@SE-FDr I added test cases and fixed another issue: The VariableSymbols need to be "deepCloned" before adding them to the scope of the OCLOperationConstraint. Otherwise, the enclosing scope of the original symbol is changed, which causes a type check error.

SE-FDr commented:
> I tried to test it, but the type cast expression is part of the UglyExprssions language which is not extended by OCL.

True, I missed that, In that case there should be only really ugly ways to do it using `let-in` expressions... Nothing that could be considered a correct solution.

```
      We add symbols from the type in 'endVisit', so we can check if there is already a
      VariableSymbol with the same name as a field form a type. In this case we do not add the
      field and users need to access it with 'this.myField'.
 ```
 I fail to see why fields get this special treatment, but functions do not (yes, function comparisons are _less_ common, but still an option).
 There is a chance we need to change this in the future to make it more consistent.
 
 The conservative solution would be to simply not support comparisons in this case, and later extend the OCL with castExpressions, but this would probably need to be discussed with Prof. Rumpe first.
Considering that methods _should_ (usually) not have the same name as variables, it probably does not matter most of the time if they are not supported at first.
@MaxStachon How relevant is this for you?

fxjordan commented:
As far as I'm concerned, we can add the same logic for functions as well. Then I'd just work around the comparison issue by using some prefixes, like:
```
class EmployeeBuilder {
    String mFirstName;
    EmployeeBuilder firstName(String pFirstName);
    Employee build();
  }
```
The only change I really need in OCL is some way to access to the methods of a type (e.g., `build()` in an operation constraint. This is not possible at the moment.

Context: I'm working on reference code adaptation of OCL artifacts and the example I showed above is one of my primary examples to show that the tool is smart enough to not produce invalid outputs like `post: build().lastName == firstName`

SE-FDr commented:
Yes, I would prefer equal treatment of functions and variables.
The rest of the MR seems fine to me.

fxjordan committed:
feat: add FunctionSymbols if there is not a local symbol yet

fxjordan commented:
Done.

One more remark: Variable/Function symbols of superclasses or even deeper nested supertypes are not accessible in the local scope of constraints/invariants yet. This is nothing I need at the moment. But maybe you want to create an internal ticket for this.
